README.md
# auth0-metrics

This library provides a module to track all frontend usage on auth0.com, it wraps Segment and sends the same data both to segment and our own endpoint. When defining new metrics, please revise the methodology at [the data playbook](https://reports.it.auth0.com/reports/data-playbook/data-collection.html#frontend-tracking).

## Installation

To install the dependencies, execute:

`npm install`



## Usage (local)
To build and run the library locally, you can run
`npm run dev`, that will let you include the library from http://localhost:9999/auth0-metrics.js, you can also test the methods included in http://localhost:9999/

## Usage (deploy)

### CDN

Using our CDN is the preferred solution. Just add the following line in your project's HTML head and you are done.

```html
<script src="https://cdn.auth0.com/website/js/1.6.0/auth0-metrics-min.js"></script>
```

Then you have to call the constructor with the correct dev/prod variables

```javascript
var metrics = new Auth0Metrics('segmentKey', 'dwhEndpoint', 'website', {});
```

#### Options {}

Additional configuration options can be specified by passing an options object to the constructor.  Metrics library will iterate over the object and add each property to the internal configuration options.

Available options:

##### removeQueryParam []{} (available in v1.6.0)

Allows stripping of query parameters from the window URL.  This is useful in cases where URL contains JWT tokens or other information that should not be recorded.  Each query parameter that needs to be removed can be specified with an object that contains a key (`string`) and a value (`RegEx string`) property.

```js
{
  removeQueryParam: [
    {
      key: 'token',
      value: '[a-z0-9\._~-]+'
    }
  ]
}
```

The following url

`http://localhost:3000/#token=xxxx.yyyy.zzzz&webtaskName=test`

will be transformed to

`http://localhost:3000/#&webtaskName=test`

### Loader

A script that will asynchronously load `auth0-metrics.js` is also provided.

```html
<script src="auth0-metrics-loader.js"></script>
<script>
  metricsLib.load({
    segmentKey: 'segmentKey',
    dwhEndpoint: 'dwhEndpoint',
    label: 'website',
    options: {}
  });
</script>
```

Once loaded, an instance of `Auth0Metrics` will be available on `metricsLib`. Calls to public instance methods of `Auth0Metrics` on `metricsLib` will be silently ignored until then.

```js
// tracks the page if auth0-metrics.js has been loaded, otherwise the call is
// ignored and doesn't produce an error.
metricsLib.page();
```

If `auth0-metrics.js` has already been loaded, the script will not attempt load it again, and will simply create an instance of `Auth0Metrics` with the given configuration.

Using the loader may also be desired when loading `auth0-metrics.js` synchronously. If there's an exception during the instantiation of `Auth0Metrics`, the stubs methods will still be in place and calls like the one above to `page` won't throw.

```html
<script src="auth0-metrics.js"></script>
<script src="auth0-metrics-loader.js"></script>
<script>
  metricsLib.load({
    segmentKey: 'segmentKey',
    dwhEndpoint: 'dwhEndpoint',
    label: 'website',
    options: {}
  });
</script>
```

Contrast the previous snippet with the following that instantiates `Auth0Metrics` directly.

```html
<script src="auth0-metrics.js"></script>
<script>
  var metricsLib = new Auth0Metrics('segmentKey', 'dwhEndpoint', 'website', {});
</script>
```

If the instantiation doesn't succeed, trying to call a method on `metricsLib` will throw because it is `undefined` and may break the behavior of your site.

## API


### .setUserId(uid)
Sets the userId of the currently connected client
#### Parameters
* `uid` user id to set

### .page(callback)
Sends information of the current page to track it
#### Parameters
* `callback` a function to call after sending this event

### .track(event, data, callback)
Sends information of a custom event to track.
#### Parameters
* `event` a string with the name of the event to track
* `data` an object with any data we need to pass
* `callback` a function to call after sending this event

### .identify(id, traits, callback)
Sends information of an identification (login/signup) to track.
> NOTE: This method's signature does not fully pair with Segment's [analytics.js](https://segment.com/docs/libraries/analytics.js/#identify). You can proxy directly to Segment by doing `metricsLib.segment().identify()` instead.

#### Parameters
* `id` user id to identify the current user to
* `traits` additional properties to set to the user
* `callback` a function to call after sending this event

You can omit the user `id` if you want to associate `traits` with the currently identified user, anonymous or not. You can also omit `traits` if all you want to do is associate an `id`. Finally, `callback` is optional and can always be omitted.

### .alias(id, callback)
Sends an alias (renaming a previous id to a new one).
#### Parameters
* `id` new user id to map to, the previous id will be taken from the cookie.
* `callback` a function to call after sending this event

### .traits()
Traits (additional properties) of the current user

### .ready(cb)
Executes a callback when segment finishes loading.
#### Parameters
* `cb` function to run when segment finishes loading.

## Issue Reporting

If you have found a bug or if you have a feature request, please report them at this repository issues section. Please do not report security vulnerabilities on the public GitHub issue tracker. The [Responsible Disclosure Program](https://auth0.com/whitehat) details the procedure for disclosing security issues.

## Author

[Auth0](auth0.com)

## License

This project is licensed under the MIT license. See the [LICENSE](LICENSE) file for more info.


CHANGELOG.md
## [2.0.0] - 2019-02-28

### Fixed

- [] Rollback change in version to use script (`Hernan Zalazar`)
  https://github.com/auth0/metrics/commit/a070127ac8f39c8baaefe439e82a09f9dee5ea10
- [] Bump version of library (`Hernan Zalazar`)
  https://github.com/auth0/metrics/commit/9a76ce4e02f10c189822bb8f6af019531d50a4d4
- [] Add `tags` option to send config for tags (`Hernan Zalazar`)
  https://github.com/auth0/metrics/commit/4648473050a5c65a40f1403ac89599963d427925
- [] Use latest auth0-tag-manager version from NPM (`Hernan Zalazar`)
  https://github.com/auth0/metrics/commit/19c691dfc6917ec4462f13d5a5af0027568c9d30
- [] Load jquery from CDN for tests (`Hernan Zalazar`)
  https://github.com/auth0/metrics/commit/8bd72338cdb833119a20842c66ee67ef46132ca8
- [] Ignore DS_Store files (`Hernan Zalazar`)
  https://github.com/auth0/metrics/commit/895c0238fe0375d8e997f67b177599174dc16135

## [1.5.0] - 2017-05-16

### Fixed

- [] Add onReady callback to loader (`Ramiro Silveyra d'Avila`)
  https://github.com/auth0/metrics/commit/ea3f426d606327d2042c90bd937dd3eaf1986e69

## [1.4.8] - 2017-05-16

### Fixed

- [] Change script to load  to metrics bundle (`Ramiro Silveyra d'Avila`)
  https://github.com/auth0/metrics/commit/bb522bb97e0e59fd9cfba7c6945f5bef15258a96

## [1.4.7] - 2017-05-16

### Fixed

- [] Inject version in loader .ejs and update cdn link to next version (`Ramiro Silveyra d'Avila`)
  https://github.com/auth0/metrics/commit/9ab75c1c730ac3f5d266e1a58f4f68c005f357e9

## [1.4.6] - 2017-05-16

### Fixed

- [] Fix loader and go back to m path in cdn (`Ramiro Silveyra d'Avila`)
  https://github.com/auth0/metrics/commit/a5dfb4d89c61f5c0693f24f444c9eb553f779cdc

## [1.4.5] - 2017-05-11

### Fixed

- [] Merge pull request #20 from mike-casas/strange-anon_id (`Miguel Casas`)
  https://github.com/auth0/metrics/commit/91d271b84c76ff20104ba2f7f042dde543e01ad9
- [] fix Strange anon_ids on memory (`Miguel`)
  https://github.com/auth0/metrics/commit/96a218cc2b97b8418724113d3b2370a31e8e9daf
- [] Fix cdn file name (`Ramiro Silveyra d'Avila`)
  https://github.com/auth0/metrics/commit/3e0cfdde70b631da49be98922df3d423448b6cf67

## [1.4.4] - 2017-05-10

### Fixed

- [] Update zuul-ngrok (`Ramiro Silveyra d'Avila`)
  https://github.com/auth0/metrics/commit/ae812f66486067307fc9f3b3284e654dd8a23d76
- [] Update cdn release process (`Ramiro Silveyra d'Avila`)
  https://github.com/auth0/metrics/commit/2fa952822fee4c6eb1b03b90012a5a4be65ba0b1
- [] Merge pull request #17 from auth0/jstrutz-patch-1 (`Alejo Fernandez`)
  https://github.com/auth0/metrics/commit/2b42fcc540b41a6f8679a53c7ca704418ba5afc1
- [] Correct release date of v1.4.3 in CHANGELOG (`Jason Strutz`)
  https://github.com/auth0/metrics/commit/29b3c44463b8a41d4cfcc4ea7917eca369e64847

## [1.4.3] - 2017-04-25

### Fixed

- [] Bump tag manager to 1.0.4 (`Alejo Fernandez`)
  https://github.com/auth0/metrics/commit/47ffd91e399bf3f539630f13568ff136bde5108f
- [] Release: 1.4.2 (`Alejo Fernandez`)
  https://github.com/auth0/metrics/commit/4d39ef6b6be1ac49bffeefccdc5504a1d0e79634
- [] Merge pull request #16 from auth0/no-segment (`Alejo Fernandez`)
  https://github.com/auth0/metrics/commit/c10eaf5d50d629898a6ad458ae2dbc5202e38307
- [] Replace segment with a custom made tag manager (`Alejo Fernandez`)
  https://github.com/auth0/metrics/commit/6286cc093f7209830da14a4d6a17f640de3683b1
- [] Release: 1.4.1 (`Cristian Douce`)
  https://github.com/auth0/metrics/commit/6690fe1fcd490c549253464ebab747a27b6bbf65
- [] Merge pull request #13 from auth0/fix/callback-not-called (`Cristian Douce`)
  https://github.com/auth0/metrics/commit/293f794a692a67f12a1dc6465d28f424d803e0b5
- [] Update index.js (`Cristian Douce`)
  https://github.com/auth0/metrics/commit/6fb7e06414912841233c191d34ae542b5596e78b
- [] Merge pull request #11 from Annyv2/license (`Nathan Totten`)
  https://github.com/auth0/metrics/commit/9f8a3049f29c042d04af84799b5754419769c9ac
- [] Update package.json (`Cristian Douce`)
  https://github.com/auth0/metrics/commit/afee19e76848b1817f5c4b38bdaddf2aa7942dbd
- [] Update README.md (`Annyv2`)
  https://github.com/auth0/metrics/commit/19203b5c91cc11fdd14003415631e1d8b9604e52
- [] Create LICENSE (`Annyv2`)
  https://github.com/auth0/metrics/commit/c004d1d36fde45b8fe49bdbdde8e46068c8ac84a

## [1.4.2] - 2015-09-11

### Fixed

- [] Merge pull request #16 from auth0/no-segment (`Alejo Fernandez`)
  https://github.com/auth0/metrics/commit/c10eaf5d50d629898a6ad458ae2dbc5202e38307
- [] Replace segment with a custom made tag manager (`Alejo Fernandez`)
  https://github.com/auth0/metrics/commit/6286cc093f7209830da14a4d6a17f640de3683b1
- [] Release: 1.4.1 (`Cristian Douce`)
  https://github.com/auth0/metrics/commit/6690fe1fcd490c549253464ebab747a27b6bbf65
- [] Merge pull request #13 from auth0/fix/callback-not-called (`Cristian Douce`)
  https://github.com/auth0/metrics/commit/293f794a692a67f12a1dc6465d28f424d803e0b5
- [] Update index.js (`Cristian Douce`)
  https://github.com/auth0/metrics/commit/6fb7e06414912841233c191d34ae542b5596e78b
- [] Merge pull request #11 from Annyv2/license (`Nathan Totten`)
  https://github.com/auth0/metrics/commit/9f8a3049f29c042d04af84799b5754419769c9ac
- [] Update package.json (`Cristian Douce`)
  https://github.com/auth0/metrics/commit/afee19e76848b1817f5c4b38bdaddf2aa7942dbd
- [] Update README.md (`Annyv2`)
  https://github.com/auth0/metrics/commit/19203b5c91cc11fdd14003415631e1d8b9604e52
- [] Create LICENSE (`Annyv2`)
  https://github.com/auth0/metrics/commit/c004d1d36fde45b8fe49bdbdde8e46068c8ac84a

## [1.4.1] - 2015-08-14

### Fixed

- [] Merge pull request #13 from auth0/fix/callback-not-called (`Cristian Douce`)
  https://github.com/auth0/metrics/commit/293f794a692a67f12a1dc6465d28f424d803e0b5
- [] Update index.js (`Cristian Douce`)
  https://github.com/auth0/metrics/commit/6fb7e06414912841233c191d34ae542b5596e78b
- [] Merge pull request #11 from Annyv2/license (`Nathan Totten`)

Gruntfile.js
var fs = require('fs');
var path = require('path');
var pkg = require('./package');

function node_bin (bin) {
  return path.join('node_modules', '.bin', bin);
}

module.exports = function (grunt) {
  grunt.initConfig({
    connect: {
      test: {
        options: {
          // base: 'test',
          hostname: '*',
          base: ['.', 'support/development-demo', 'support/development-demo/build', 'build'],
          port: 9999
        }
      },
      demo: {
        options: {
          hostname: '*',
          base: ['support/development-demo', 'support/development-demo/build', 'build', '.'],
          port: 3000
        }
      }
    },
    browserify: {
      options: {
        browserifyOptions: {
          debug: true
        },
        watch: true,

        // Convert absolute sourcemap filepaths to relative ones using mold-source-map.
        postBundleCB: function(err, src, cb) {
          if (err) { return cb(err); }
          var through = require('through');
          var stream = through().pause().queue(src).end();
          var buffer = '';

          stream.pipe(require('mold-source-map').transformSourcesRelativeTo(__dirname)).pipe(through(function(chunk) {
            buffer += chunk.toString();
          }, function() {
            cb(err, buffer);
          }));
          stream.resume();
        }

      },
      debug: {
        files: {
          'build/auth0-metrics.js': ['standalone.js']
        }
      },
    },
    less: {
      demo: {
        files: {
          'support/development-demo/build/index.css': 'support/development-demo/index.less'
        }
      }
    },
    copy: {
      demo: {
        files: {
          'support/development-demo/auth0-metrics.min.js': 'build/auth0-metrics.min.js',
          'support/development-demo/auth0-metrics.js':     'build/auth0-metrics.js'
        }
      }
    },
    exec: {
      'uglify': {
        cmd: node_bin('uglifyjs') + ' build/auth0-metrics.js  -b beautify=false,ascii_only=true > build/auth0-metrics.min.js',
        stdout: true,
        stderr: true
      },
      'uglify-loader': {
        cmd: node_bin('uglifyjs') + ' build/auth0-metrics-loader.js  -b beautify=false,ascii_only=true > build/auth0-metrics-loader.min.js',
        stdout: true,
        stderr: true
      },
      'test-inception': {
        cmd: node_bin('mocha') + ' ./test/support/characters-inception.test.js',
        stdout: true,
        stderr: true
      },
      'test-integration': {
        cmd: node_bin('zuul') + ' -- test/*.js',
        stdout: true,
        stderr: true
      },
      'test-phantom': {
        cmd: node_bin('zuul') + ' --ui mocha-bdd --disable-tunnel --phantom 9999 -- test/*.js',
        stdout: true,
        stderr: true
      },
      cdn: {
        cmd: node_bin('ccu'),
        stdout: true,
        stderr: true
      }
    },
    clean: {
      js: ['release/', 'build/', 'support/development-demo/auth0-metrics.js']
    },
    watch: {
      js: {
        files: ['build/auth0-metrics.js'],
        tasks: [],
        options: {
          livereload: true
        },
      },
      demo: {
        files: ['support/development-demo/*'],
        tasks: ['less:demo'],
        options: {
          livereload: true
        },
      }
    },
    jsdoc: {
      dist : {
        src: 'index.js',
        options: {
          destination: 'build',
          template : "support/loader",
          query: "version=" + pkg.version
        }
      }
    }
  });


  // Loading dependencies
  for (var key in grunt.file.readJSON('package.json').devDependencies) {
    if (key !== 'grunt' && key.indexOf('grunt') === 0) { grunt.loadNpmTasks(key); }
  }

  grunt.registerTask('js',            ['clean:js', 'browserify:debug', 'exec:uglify']);
  grunt.registerTask('loader',        ['jsdoc', 'exec:uglify-loader']);
  grunt.registerTask('build',         ['js', 'loader']);

  grunt.registerTask('demo',          ['less:demo', 'connect:demo', 'build', 'watch']);

  grunt.registerTask('dev',           ['connect:test', 'build', 'watch']);
  grunt.registerTask('integration',   ['exec:test-inception', 'exec:test-integration']);
  grunt.registerTask('phantom',       ['build', 'exec:test-inception', 'exec:test-phantom']);

  grunt.registerTask('cdn',           ['build', 'exec:cdn']);
};
